File: apps/api/src/lib/mailer.ts
import { Resend } from 'resend';
import { env } from '../env.js';

export type Mail = {
  to: string;
  subject: string;
  html: string;
  /** Rendered inline in the HTML as a data URL and attached as a file, so the
   *  ticket survives a client that strips inline images. */
  attachments?: { filename: string; content: Buffer }[];
};

const resend = env.RESEND_API_KEY ? new Resend(env.RESEND_API_KEY) : null;

/**
 * Sends, or explains loudly why it did not.
 *
 * With no API key configured the message is logged rather than dropped, so a
 * fresh clone can walk the whole booking flow without an email account. That
 * is a development convenience, never a silent production failure — the log
 * says exactly what would have been sent and to whom.
 */
export async function sendMail(mail: Mail): Promise<void> {
  if (!resend) {
    console.warn(`[mail] RESEND_API_KEY not set — would have sent "${mail.subject}" to ${mail.to}`);
    return;
  }

  // Production never redirects: a customer's ticket must reach the customer.
  const redirect = env.NODE_ENV === 'production' ? undefined : env.MAIL_REDIRECT_TO;
  const to = redirect ?? mail.to;
  const subject = redirect ? `[to: ${mail.to}] ${mail.subject}` : mail.subject;

  if (redirect) console.log(`[mail] redirected ${mail.to} -> ${redirect}`);

  const { error } = await resend.emails.send({
    from: env.MAIL_FROM,
    to,
    subject,
    html: mail.html,
    ...(mail.attachments ? { attachments: mail.attachments } : {}),
  });

  // Thrown, not swallowed: this runs inside a job with retry and backoff, and
  // a failure that returns quietly would burn the retry budget doing nothing.
  if (error) {
    throw new Error(`Resend refused the message: ${error.name} — ${error.message}`);
  }
}

File: apps/api-py/src/ticket_api/lib/mailer.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass, field

import resend

from ..config import IS_PROD, settings


@dataclass(slots=True)
class Mail:
    to: str
    subject: str
    html: str
    #: Rendered inline in the HTML as a data URL *and* attached as a file, so
    #: the ticket survives a client that strips inline images.
    attachments: list[dict[str, object]] = field(default_factory=list)


if settings.RESEND_API_KEY:
    resend.api_key = settings.RESEND_API_KEY


async def send_mail(mail: Mail) -> None:
    """
    Sends, or explains loudly why it did not.

    With no API key configured the message is logged rather than dropped, so a
    fresh clone can walk the whole booking flow without an email account. That
    is a development convenience, never a silent production failure — the log
    says exactly what would have been sent and to whom.
    """
    if not settings.RESEND_API_KEY:
        print(f'[mail] RESEND_API_KEY not set — would have sent "{mail.subject}" to {mail.to}')
        return

    # Production never redirects: a customer's ticket must reach the customer.
    redirect = None if IS_PROD else settings.MAIL_REDIRECT_TO
    to = redirect or mail.to
    subject = f"[to: {mail.to}] {mail.subject}" if redirect else mail.subject

    if redirect:
        print(f"[mail] redirected {mail.to} -> {redirect}")

    params: dict[str, object] = {
        "from": settings.MAIL_FROM,
        "to": [to],
        "subject": subject,
        "html": mail.html,
    }
    if mail.attachments:
        params["attachments"] = mail.attachments

    # The Resend SDK is synchronous; run it off the event loop so a slow
    # provider cannot stall every other coroutine in the worker.
    #
    # Raised, not swallowed: this runs inside a job with retry and backoff, and
    # a failure that returns quietly would burn the retry budget doing nothing.
    await asyncio.to_thread(resend.Emails.send, params)
